packages/http/src/forwarder/HttpForwarder.ts

```diff
@@ -7,21 +7,17 @@ export class HttpForwarder
   implements IForwarder<IHttpOperation, IHttpRequest, IHttpConfig, IHttpResponse> {
   public async forward(opts: {
     resource?: IHttpOperation;
-    // todo: IHttpRequest should be enough
     input: IPrismInput<IHttpRequest>;
   }): Promise<IHttpResponse> {
-    if (!opts.resource) {
-      throw new Error('Missing spec');
-    }
-    if (!opts.resource.servers || opts.resource.servers.length === 0) {
-      throw new Error('Server list is missing in spec');
-    }
-
     const inputData = opts.input.data;
 
     const response = await axios({
       method: inputData.method,
-      url: this.resolveServerUrl(opts.resource.servers[0]) + inputData.url.path,
+      baseURL:
+        opts.resource && opts.resource.servers && opts.resource.servers.length > 0
+          ? this.resolveServerUrl(opts.resource.servers[0])
+          : inputData.url.baseUrl,
+      url: inputData.url.path,
```

Comment:
This looks good but you'll need to add some integration tests to:

- packages/http/src/__tests__/http-prism-instance.spec.ts
- packages/http-server/src/__tests__/server.spec.ts

Verify that you can "have no resources, yet still be able to run prism and validation and routing doesn't fail". This is critical because `prism.process` will do routing and validation first (which rely on resources) and forwarding later.

Reply:
That makes a lot of sense! I'll add the tests!

Reply:
@chris-miaskowski The routing will fail since it seems like it requires resources to make it work properly, and I guess it's fine, isn't it?

Reply:
@chris-miaskowski I'm also not so sure what I should be testing in the second file — since having a server without a document is not something we're aiming at, right?

Reply:
@XVincentX 
> The routing will fail since it seems like it requires resources to make it work properly.

That's what I was worried about. If `router` throws then `forwarder` won't even be executed.

But the question is **"what are you planning to use in you request maker eventually? `prism.process` or the `forwarded.forward`"**?

If the former then you must change your implementation (e.g. allow prism.process to take a special config parameter that allows no-resources), if the former then your tests are fine.

There is another problem I have realised, though.

What if someone uses `prism.process` but `router.router` returns empty array AND inputData.url.baseUrl OR inputData.url.path is undefined?

Reply:
> That's what I was worried about. If `router` throws then `forwarder` won't even be executed.

That is correct; however the router can be skipped by providing an undefined instance when creating the Prism instance: https://github.com/stoplightio/prism/blob/feat/forwarder-fix/packages/core/src/factory.ts#L38

> If the former then you must change your implementation (e.g. allow prism.process to take a special config parameter that allows no-resources), if the former then your tests are fine.

I believe that right now both would work. I can either use the forwarder directly or call the process method: as long I do not provide the router component, I can still use that function call.

> What if someone uses `prism.process` but `router.router` returns empty array AND inputData.url.baseUrl OR inputData.url.path is undefined?

You get an exception, and you kind of deserved it I guess.

Reply:
> the router can be skipped by providing an undefined instance when creating the Prism instance

Yes but I wouldn't want Prism to rely on this sort of hacks and insider knowledge. Prism must prevent this situation from happening either by exposing the right interface or by throwing errors.

In other words forwarder should not be dependent on a combination of "you must provide router and resources, if you don't provide resources you should not provide router". This is just too complex.

> I believe that right now both would work. I can either use the forwarder directly or call the process method: as long I do not provide the router component, I can still use that function call.

If you want to use `forwarder` then go for it, but using using `prism.process` and relying on the complex condition "don't provide router if you don't provide resource" is a bad idea.

My suggestion is:
- either use forwarder
- or add explicit `config.allowNoResources` (or `config.forwardOnly`) prop and make sure router and validator aren't invoked.

> You get an exception, and you kind of deserved it I guess.

What exception? As far as I can see we don't throw any explicit exception.

Reply:
> Yes but I wouldn't want Prism to rely on this sort of hacks and insider knowledge

> In other words forwarder should not be dependent on a combination of "you must provide router and resources, if you don't provide resources you should not provide router". This is just too complex.

If you think that is a hack, then there's a design problem. 

<img width="761" alt="image" src="https://user-images.githubusercontent.com/1416224/52639183-05fe3100-2ed4-11e9-8f93-26f58f376177.png">

That insider knowledge is exposed as a top level property that can even be undefined so — it is a possibility. If you do not want that, then the library should not allow that behavior at all.

---

That said, I vote to use the `forwarder`, I'm ok with that.